#### 과제:
p496까지 예제풀기

In [1]:
!pip install nltk

In [2]:
from nltk import sent_tokenize #nltk : 방대한 데이터 세트와 서브모듈, 다양한 데이터 세트를 지원(가장 보편적)
import nltk

# nltk.download('punkt')
text_sample = 'The Matrix is everywhere its all around us, here even in this room. \
                        You can see it out your window or on your television. \
                        You feel it when you do to work, or go to church or pay your taxes.'
sentences = sent_tokenize(text = text_sample)

# Stemming 과 Lemmatization

1) Stemming

In [3]:
from nltk.stem import LancasterStemmer
stemmer = LancasterStemmer()

print(stemmer.stem('working'), stemmer.stem('works'), stemmer.stem('worked'))
print(stemmer.stem('happier'), stemmer.stem('happiest'), stemmer.stem('happinist'))
print(stemmer.stem('amusing'), stemmer.stem('amuses'), stemmer.stem('amused'))

work work work
happy happiest happin
amus amus amus


-> work 말고는 원형 추출을 잘 못함(품사 인식을 못함)

2) Lemmatization

In [4]:
from nltk.stem import WordNetLemmatizer
import nltk
nltk.download('wordnet')
nltk.download('omw-1.4')

lemma = WordNetLemmatizer()

print(lemma.lemmatize('amusing', 'v'), lemma.lemmatize('amuses', 'v'), lemma.lemmatize('amused', 'v'))
print(lemma.lemmatize('happier', 'a'), lemma.lemmatize('happiest', 'a'), lemma.lemmatize('happinist', 'n'))

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


amuse amuse amuse
happy happy happinist


# 희소행렬

## COO 형식

In [5]:
from scipy import sparse
import numpy as np

# 데이터
data = np.array([3, 1, 2]) # 0이 아닌 데이터

# 각 데이터의 행과 열의 위치
row_pos = np.array([0, 0, 1])
col_pos = np.array([0, 2, 1])

# sparse 패키지의 coo_matrix 이용 COO 형식으로 희소 행렬 생성
sparse_coo = sparse.coo_matrix((data, (row_pos, col_pos)))
print(sparse_coo)

# 다시 밀집형태의 행렬로 출력
sparse_coo.toarray()

  (0, 0)	3
  (0, 2)	1
  (1, 1)	2


array([[3, 0, 1],
       [0, 2, 0]])

In [6]:
dense2 = np.array([[0, 0, 1, 0, 0, 5], 
                               [1, 4, 0, 3, 2, 5],
                               [0, 6, 0, 3, 0, 0],
                               [2, 0, 0, 0, 0, 0],
                               [0, 0, 0, 7, 0, 8],
                               [1, 0, 0, 0, 0, 0]])

# 0이 아닌 데이터 추출
data2 = np.array([1, 5, 1, 4, 3, 2, 5, 6, 3, 2, 7, 8, 1])

row_pos = np.array([0, 0, 1, 1, 1, 1, 1, 2, 2, 3, 4, 4, 5])
col_pos = np.array([2, 5, 0, 1, 3, 4, 5, 1, 3, 0, 3, 5, 0])

spars_coo = sparse.coo_matrix((data2, (row_pos, col_pos)))
print(spars_coo, '\n')
print('밀집형태')
spars_coo.toarray()

  (0, 2)	1
  (0, 5)	5
  (1, 0)	1
  (1, 1)	4
  (1, 3)	3
  (1, 4)	2
  (1, 5)	5
  (2, 1)	6
  (2, 3)	3
  (3, 0)	2
  (4, 3)	7
  (4, 5)	8
  (5, 0)	1 

밀집형태


array([[0, 0, 1, 0, 0, 5],
       [1, 4, 0, 3, 2, 5],
       [0, 6, 0, 3, 0, 0],
       [2, 0, 0, 0, 0, 0],
       [0, 0, 0, 7, 0, 8],
       [1, 0, 0, 0, 0, 0]])

In [7]:
# CSR
# 행위치 배열의 고유한 값들의 시작 위치 인덱스를 배열로 생성
row_pos_ind = np.array([0, 2, 7, 9, 10, 12, 13])

sparse_csr = sparse.csr_matrix((data2, col_pos, row_pos_ind))
print(sparse_csr, '\n')
print('CSR 변환된 데이터 다시 Dense로 출력 확인')
sparse_csr.toarray()

  (0, 2)	1
  (0, 5)	5
  (1, 0)	1
  (1, 1)	4
  (1, 3)	3
  (1, 4)	2
  (1, 5)	5
  (2, 1)	6
  (2, 3)	3
  (3, 0)	2
  (4, 3)	7
  (4, 5)	8
  (5, 0)	1 

CSR 변환된 데이터 다시 Dense로 출력 확인


array([[0, 0, 1, 0, 0, 5],
       [1, 4, 0, 3, 2, 5],
       [0, 6, 0, 3, 0, 0],
       [2, 0, 0, 0, 0, 0],
       [0, 0, 0, 7, 0, 8],
       [1, 0, 0, 0, 0, 0]])

#### 과제 1019_1. 10행 10열 희소행렬(0의 비중을 70% 이상)을 생성한 후 COO, CSR 방식으로 변환 후 다시 희소행렬로 출력

In [90]:
array = np.zeros((10, 10))
print(array)

[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]


In [91]:
# 나머지 다른 숫자로 채워주기
# 열부분
row = []
for ind in range(4):
    row.append(np.random.randint(1, 10))

# 행부분
col = []
for co in range(10):
    col.append(np.random.randint(1, 10))
    
print(row)
print(col)

for ro in row:
    for co in col:
        array[co, ro] = np.random.randint(1, 10)
array        

[2, 2, 5, 5]
[9, 4, 9, 6, 1, 7, 6, 2, 2, 6]


array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 7., 0., 0., 0., 0.],
       [0., 0., 4., 0., 0., 5., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 5., 0., 0., 4., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 2., 0., 0., 3., 0., 0., 0., 0.],
       [0., 0., 8., 0., 0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 3., 0., 0., 6., 0., 0., 0., 0.]])

## 1) COO 방식으로 변환

In [102]:
from scipy import sparse

# 0이 아닌 데이터의 위치 
row_pos = []
col_pos = []
data_ = []

for co, arr in enumerate(array):
    for ro, num in enumerate(arr):
        if num != 0:
            row_pos.append(ro)
            col_pos.append(co)
            data_.append(num)
            
print(col_pos)
print(row_pos)
print(data_)

# sparse 패키지의 coo_matrix 이용
sparse_coo = sparse.coo_matrix((data_, (np.array(col_pos), np.array(row_pos))))

[1, 1, 2, 2, 4, 4, 6, 6, 7, 7, 9, 9]
[2, 5, 2, 5, 2, 5, 2, 5, 2, 5, 2, 5]
[1.0, 7.0, 4.0, 5.0, 5.0, 4.0, 2.0, 3.0, 8.0, 1.0, 3.0, 6.0]


In [103]:
# 다시 밀집 형태의 행렬로 출력
sparse_coo.toarray()

array([[0., 0., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 7.],
       [0., 0., 4., 0., 0., 5.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 5., 0., 0., 4.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 2., 0., 0., 3.],
       [0., 0., 8., 0., 0., 1.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 3., 0., 0., 6.]])

## 2) CSR 방식

In [111]:
array

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 7., 0., 0., 0., 0.],
       [0., 0., 4., 0., 0., 5., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 5., 0., 0., 4., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 2., 0., 0., 3., 0., 0., 0., 0.],
       [0., 0., 8., 0., 0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 3., 0., 0., 6., 0., 0., 0., 0.]])

In [115]:
# 행 위치 배열의 고유 값 시작위치 인덱스 저장

# 행 위치 인덱스
print(np.array(col_pos))

csr_col = []
fir_arr = 0
for ind, num in enumerate(col_pos):
    if fir_arr != num:
        csr_col.append(ind)
        fir_arr = num
    else: pass

print(csr_col)
# 마지막 행 추가
csr_col.append(11)
csr_col

[1 1 2 2 4 4 6 6 7 7 9 9]
[0, 2, 4, 6, 8, 10]


[0, 2, 4, 6, 8, 10, 11]

In [116]:
# csr 형식으로 변환
sparse_csr = sparse.csr_matrix((data_, np.array(row_pos), np.array(csr_col)))

# CSR 다시 밀집 형태의 행렬로 출력
sparse_csr.toarray()

array([[0., 0., 1., 0., 0., 7.],
       [0., 0., 4., 0., 0., 5.],
       [0., 0., 5., 0., 0., 4.],
       [0., 0., 2., 0., 0., 3.],
       [0., 0., 8., 0., 0., 1.],
       [0., 0., 3., 0., 0., 0.]])

# 텍스트 분류 실습 - 20 뉴스그룹 분류

In [8]:
from sklearn.datasets import fetch_20newsgroups

news_data = fetch_20newsgroups(subset = 'all', random_state = 156)

print(news_data.keys())

dict_keys(['data', 'filenames', 'target_names', 'target', 'DESCR'])


-> filenames : API가 인터넷에서 내려받아 로컬 컴퓨터에 저장하는 디렉터리와 파일명

## 1. 데이터 확인

target 클래스 구성 확인

In [9]:
import pandas as pd

print('Target 클래스의 값과 분포도')
print(pd.Series(news_data.target).value_counts().sort_index(), '\n')
print('Target 클래스의 이름')
news_data.target_names

Target 클래스의 값과 분포도
0     799
1     973
2     985
3     982
4     963
5     988
6     975
7     990
8     996
9     994
10    999
11    991
12    984
13    990
14    987
15    997
16    910
17    940
18    775
19    628
dtype: int64 

Target 클래스의 이름


['alt.atheism',
 'comp.graphics',
 'comp.os.ms-windows.misc',
 'comp.sys.ibm.pc.hardware',
 'comp.sys.mac.hardware',
 'comp.windows.x',
 'misc.forsale',
 'rec.autos',
 'rec.motorcycles',
 'rec.sport.baseball',
 'rec.sport.hockey',
 'sci.crypt',
 'sci.electronics',
 'sci.med',
 'sci.space',
 'soc.religion.christian',
 'talk.politics.guns',
 'talk.politics.mideast',
 'talk.politics.misc',
 'talk.religion.misc']

data 구성 확인

In [10]:
news_data.data[0]

'From: egreen@east.sun.com (Ed Green - Pixel Cruncher)\nSubject: Re: Observation re: helmets\nOrganization: Sun Microsystems, RTP, NC\nLines: 21\nDistribution: world\nReply-To: egreen@east.sun.com\nNNTP-Posting-Host: laser.east.sun.com\n\nIn article 211353@mavenry.altcit.eskimo.com, maven@mavenry.altcit.eskimo.com (Norman Hamer) writes:\n> \n> The question for the day is re: passenger helmets, if you don\'t know for \n>certain who\'s gonna ride with you (like say you meet them at a .... church \n>meeting, yeah, that\'s the ticket)... What are some guidelines? Should I just \n>pick up another shoei in my size to have a backup helmet (XL), or should I \n>maybe get an inexpensive one of a smaller size to accomodate my likely \n>passenger? \n\nIf your primary concern is protecting the passenger in the event of a\ncrash, have him or her fitted for a helmet that is their size.  If your\nprimary concern is complying with stupid helmet laws, carry a real big\nspare (you can put a big or small 

- 뉴스그룹 제목, 작성자, 소속, 이메일 등 다양한 정보 포함

-> 뉴스 내용을 제외한 나머지 정보는 제거 (Target 클래스 값과 유사한 데이터를 가지고 있는 경우가 많아서)

In [11]:
## 

In [12]:
# subset = train 으로 학습용 데이터만 추출
train_news = fetch_20newsgroups(subset = 'train', remove = ('headers', 'footers', 'quotes'), random_state = 156)
X_train = train_news.data
y_train = train_news.target

# subset = test 으로 테스트 데이터만 추출
test_news = fetch_20newsgroups(subset = 'test', remove = ('headers', 'footers', 'quotes'), random_state = 156)

X_test = test_news.data
y_test = test_news.target
print(len(train_news.data), len(test_news.data))

11314 7532


## 피처 벡터화 변환

### 1) Count 기반의 피처 벡터화

In [13]:
from sklearn.feature_extraction.text import CountVectorizer

# Count Vectorization으로 피처 벡터화 변환 수행
cnt_vect = CountVectorizer()
cnt_vect.fit(X_train)
X_train_cnt_vect = cnt_vect.transform(X_train) 

# 학습데이터로 fit 된 CountVectorizer를 이용해 텍스트를 피처 벡터화 변환 수행 (왜???_훈련용과 test 용의 피처를 동일하게 하기위해)
X_test_cnt_vect = cnt_vect.transform(X_test) 

print(X_train_cnt_vect.shape)
print(X_test_cnt_vect.shape)

(11314, 101631)
(7532, 101631)


-> 컬럼수가 많이 늘어남

### 2) TF-IDF기반의 피처 벡터화

In [14]:
from sklearn.feature_extraction.text import TfidfVectorizer

# TF-IDF 벡터화
tfidf_vect = TfidfVectorizer()
tfidf_vect.fit(X_train)

X_train_tfidf_vect = tfidf_vect.transform(X_train)
X_test_tfidf_vect = tfidf_vect.transform(X_test)

print(X_train_tfidf_vect.shape)
print(X_test_tfidf_vect.shape)

(11314, 101631)
(7532, 101631)


## 머신러닝 모델 학습/예습/평가

### 1) Count기반의 피처 벡터화 데이터

In [15]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import warnings
warnings.filterwarnings('ignore')

lr_clf = LogisticRegression(solver = 'liblinear')
lr_clf.fit(X_train_cnt_vect, y_train)
pred = lr_clf.predict(X_test_cnt_vect)
print('정확도: ', accuracy_score(y_test, pred))

정확도:  0.6167020711630377


### 2) TF-IDF기반의 피처 벡터화 데이터

In [16]:
lr_clf = LogisticRegression(solver = 'liblinear')
lr_clf.fit(X_train_tfidf_vect, y_train)
pred_td = lr_clf.predict(X_test_tfidf_vect)

print('TF-IDF 정확도: ', accuracy_score(y_test, pred_td))

TF-IDF 정확도:  0.6775092936802974


-> TF-IDF가 단순 Count 기반보다 훨씬 높은 예측 정확도 제공

- 일반적으로 문서 내에 텍스트가 많고 많은 문서를 가지는 텍스트 분석에서 카운트 벡터화보다는 TF-IDF벡터화가 좋은 예측결과 도출

텍스트 분석에서 머신러인 모델의 성능을 향상시키는 중요한 2가지 방법
- 1) 최적의 ml 알고리즘을 선택
- 2) 최상의 피처 전처리를 수행

#### 과제 1019_2. TfidfVectorizer 클래스의 stopword를 'english'로 변경, ngram_range는 (1, 2)로 max_df = 300 으로 변경후 다시 예측 성능을 수행

## TF-IDF 벡터화의 파라미터 변경 적용

In [17]:
# stop words 필터링 추가, ngram : (1,1) -> (1, 2) 변경, max-df = 300 로 변경

tfidf_vect = TfidfVectorizer(stop_words='english', ngram_range = (1,2), max_df = 300)
tfidf_vect.fit(X_train)
X_train_tfidf_vect = tfidf_vect.transform(X_train)
X_test_tfidf_vect = tfidf_vect.transform(X_test)

lr_clf = LogisticRegression(solver= 'liblinear')
lr_clf.fit(X_train_tfidf_vect, y_train)
pred = lr_clf.predict(X_test_tfidf_vect)

print('TF-IDF 벡터화의 파라미터 변경 된 예측 정확도: ', accuracy_score(y_test, pred))

TF-IDF 벡터화의 파라미터 변경 된 예측 정확도:  0.6901221455124801


-> 단순 파라미터보다 더 향상 됨 (2% 향상)

#### 과제 1019_3. GridSearchCV를 이용해 하이퍼파라미터 튜닝을 수행하여 예측 성능을 개선

## GridSearchCV를 이용한 로지스틱 회귀의 하이퍼 파라미터 최적화 수행

In [18]:
from sklearn.model_selection import GridSearchCV

# 최적 C값 도출 튜닝 수행. CV는 3폴드 세트로 설정

params = {'C' : [0.01, 0.1, 1, 5, 10]}
grid_cv_lr = GridSearchCV(lr_clf, param_grid = params, cv = 3, scoring = 'accuracy', verbose = 1)
grid_cv_lr.fit(X_train_tfidf_vect, y_train)
print('로지스틱 회귀 best C parameter :', grid_cv_lr.best_params_)

# 최적 C값으로 학습된 grid_cv로 예측 및 정확도 평가
pred = grid_cv_lr.predict(X_test_tfidf_vect)
print('TF-IDF 벡터화의 로지스틱 예측 정확도 :', accuracy_score(y_test, pred))

Fitting 3 folds for each of 5 candidates, totalling 15 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  15 out of  15 | elapsed:  2.3min finished


로지스틱 회귀 best C parameter : {'C': 10}
TF-IDF 벡터화의 로지스틱 예측 정확도 : 0.7039298990971854


-> C (알파의 역수) 의 최적의 파라미터가 10 : 높을 수록 규제가 약해짐

[Count 벡터화]
예측 정확도 : 0.6167
    
[단순 TF-IDF 벡터화]
예측 정확도 : 0.677

[TF-IDF 벡터화의 파라미터 변경 적용]
예측 정확도 : 0.690

# 감성 분석

## 1. 지도학습 - [실습] IMDB 영화평
- 영화평의 텍스트를 분석해 감성 분석 결과가 긍정 or 부정인지 예측하는 모델 생성

In [19]:
review_df = pd.read_csv('labeledTrainData.tsv', header=0, sep = '\t', quoting = 3)
review_df.head()

id  sentiment                                             review
0  "5814_8"          1  "With all this stuff going down at the moment ...
1  "2381_9"          1  "\"The Classic War of the Worlds\" by Timothy ...
2  "7759_3"          0  "The film starts with a manager (Nicholas Bell...
3  "3630_4"          0  "It must be assumed that those who praised thi...
4  "9495_8"          1  "Superbly trashy and wondrously unpretentious ...

id : 각 데이터의 id

sentiment : 영화평의 Sentiment 결과 값(Target Label) 1은 긍정적 평가, 0은 부정적 평가

review : 영화평의 텍스트

In [20]:
review_df['review'][0]

'"With all this stuff going down at the moment with MJ i\'ve started listening to his music, watching the odd documentary here and there, watched The Wiz and watched Moonwalker again. Maybe i just want to get a certain insight into this guy who i thought was really cool in the eighties just to maybe make up my mind whether he is guilty or innocent. Moonwalker is part biography, part feature film which i remember going to see at the cinema when it was originally released. Some of it has subtle messages about MJ\'s feeling towards the press and also the obvious message of drugs are bad m\'kay.<br /><br />Visually impressive but of course this is all about Michael Jackson so unless you remotely like MJ in anyway then you are going to hate this and find it boring. Some may call MJ an egotist for consenting to the making of this movie BUT MJ and most of his fans would say that he made it for the fans which if true is really nice of him.<br /><br />The actual feature film bit when it finally

"< br >" 태그 삭제 필요, 숫자/특수문자 변경 필요

In [22]:
import re

# < br >태그 공백 전환
review_df['review'] = review_df['review'].str.replace('<br />', ' ')

# 영어 문자열이 아닌 문자 전부 공백으로 전환
review_df['review'] = review_df['review'].apply(lambda x :re.sub("[^a-zA-Z]", " ", x))

review_df['review'][0]

' With all this stuff going down at the moment with MJ i ve started listening to his music  watching the odd documentary here and there  watched The Wiz and watched Moonwalker again  Maybe i just want to get a certain insight into this guy who i thought was really cool in the eighties just to maybe make up my mind whether he is guilty or innocent  Moonwalker is part biography  part feature film which i remember going to see at the cinema when it was originally released  Some of it has subtle messages about MJ s feeling towards the press and also the obvious message of drugs are bad m kay   Visually impressive but of course this is all about Michael Jackson so unless you remotely like MJ in anyway then you are going to hate this and find it boring  Some may call MJ an egotist for consenting to the making of this movie BUT MJ and most of his fans would say that he made it for the fans which if true is really nice of him   The actual feature film bit when it finally starts is only on for 

## 데이터 분류

In [24]:
# 결정 값 클래스인 sentiment 칼럼을 Target 값 데이터 세트로 생성
# 피처 데이터 : id, sentiment 칼럼을 삭제 후 생성

from sklearn.model_selection import train_test_split

class_df = review_df['sentiment']
feature_df = review_df.drop(['id', 'sentiment'], axis = 1, inplace = False)

X_train, X_test, y_train, y_test = train_test_split(feature_df, class_df, test_size = 0.3, random_state = 156)
X_train.shape, X_test.shape

((17500, 1), (7500, 1))

## review 텍스트 피처 벡터화 및 모델 훈련/예측/평가

### 1) Count 벡터화

In [27]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, roc_auc_score

# Pipeline 객체 이용 (피처 벡터화, ML 분류 알고리즘 생성)
# 스톱 워드 Emglish, filtering, ngram 은 (1, 2) 설정 후 CountVectorization 수행
# 로지스틱 C는 10으로 설정

pipeline = Pipeline([
    ('cnt_vect', CountVectorizer(stop_words='english', ngram_range=(1, 2) )),
    ('lr_clf', LogisticRegression(solver = 'liblinear', C = 10) )])

# 모델 학습, 예측, 수행
pipeline.fit(X_train['review'], y_train)
pred = pipeline.predict(X_test['review'])
pred_probs = pipeline.predict_proba(X_test['review'])[:, 1]

print('예측 정확도 :', accuracy_score(y_test, pred))
print('ROC_AUC :', roc_auc_score(y_test, pred_probs))

예측 정확도 : 0.886
ROC_AUC : 0.9502646966765308


#### 과제 1019_4. TF-IDF 벡터화를 적용하여 수행(스톱워드, ngram을 추가)

### 2) TF-IDF 벡터화

In [28]:
# Pipeline 객체 이용 (피처 벡터화, ML 분류 알고리즘 생성)
# 스톱 워드 Emglish, filtering, ngram 은 (1, 2) 설정 후 CountVectorization 수행
# 로지스틱 C는 10으로 설정

pipeline = Pipeline([
    ('cnt_vect', TfidfVectorizer(stop_words='english', ngram_range=(1, 2) )),
    ('lr_clf', LogisticRegression(solver = 'liblinear', C = 10) )])

# 모델 학습, 예측, 수행
pipeline.fit(X_train['review'], y_train)
pred = pipeline.predict(X_test['review'])
pred_probs = pipeline.predict_proba(X_test['review'])[:, 1]

print('TF-IDF 벡터화 예측 정확도 :', accuracy_score(y_test, pred))
print('TF-IDF 벡터화 ROC_AUC :', roc_auc_score(y_test, pred_probs))

TF-IDF 벡터화 예측 정확도 : 0.8936
TF-IDF 벡터화 ROC_AUC : 0.959800534941953


-> TF_IDF 기반 피처 벡터화의 예측 성능이 조금 더 나아짐

## 2. 비지도학습

In [30]:
import nltk
nltk.download('all')

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to
[nltk_data]    |     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping corpora\abc.zip.
[nltk_data]    | Downloading package alpino to
[nltk_data]    |     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping corpora\alpino.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger to
[nltk_data]    |     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping taggers\averaged_perceptron_tagger.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger_ru to
[nltk_data]    |     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping
[nltk_data]    |       taggers\averaged_perceptron_tagger_ru.zip.
[nltk_data]    | Downloading package basque_grammars to
[nltk_data]    |     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping grammars\basque_grammars.zip.
[nltk_data]   

[nltk_data]    | Downloading package nonbreaking_prefixes to
[nltk_data]    |     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping corpora\nonbreaking_prefixes.zip.
[nltk_data]    | Downloading package nps_chat to
[nltk_data]    |     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping corpora\nps_chat.zip.
[nltk_data]    | Downloading package omw to
[nltk_data]    |     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]    | Downloading package omw-1.4 to
[nltk_data]    |     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]    |   Package omw-1.4 is already up-to-date!
[nltk_data]    | Downloading package opinion_lexicon to
[nltk_data]    |     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping corpora\opinion_lexicon.zip.
[nltk_data]    | Downloading package panlex_swadesh to
[nltk_data]    |     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]    | Downloading package paradigms to
[nltk_data]    | 

[nltk_data]    |   Unzipping corpora\verbnet3.zip.
[nltk_data]    | Downloading package webtext to
[nltk_data]    |     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping corpora\webtext.zip.
[nltk_data]    | Downloading package wmt15_eval to
[nltk_data]    |     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping models\wmt15_eval.zip.
[nltk_data]    | Downloading package word2vec_sample to
[nltk_data]    |     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping models\word2vec_sample.zip.
[nltk_data]    | Downloading package wordnet to
[nltk_data]    |     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]    |   Package wordnet is already up-to-date!
[nltk_data]    | Downloading package wordnet2021 to
[nltk_data]    |     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]    | Downloading package wordnet31 to
[nltk_data]    |     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]    | Downloading package 

True

### 1) SentiWordNet 이용한 감정 분석

In [31]:
from nltk.corpus import wordnet as wn

term = 'present'

# present 라는 단어로 wordnet 의 synsets생성
synsets = wn.synsets(term)
print('synsets 반환 type: ', type(synsets))
print('synsets 반환 값 개수: ', len(synsets))
print('synsets 반환 값: ', synsets)

synsets 반환 type:  <class 'list'>
synsets 반환 값 개수:  18
synsets 반환 값:  [Synset('present.n.01'), Synset('present.n.02'), Synset('present.n.03'), Synset('show.v.01'), Synset('present.v.02'), Synset('stage.v.01'), Synset('present.v.04'), Synset('present.v.05'), Synset('award.v.01'), Synset('give.v.08'), Synset('deliver.v.01'), Synset('introduce.v.01'), Synset('portray.v.04'), Synset('confront.v.03'), Synset('present.v.12'), Synset('salute.v.06'), Synset('present.a.01'), Synset('present.a.02')]


- 서로 다른 총 18개의 semantic을 가지는 synset 객체를 가지는 리스트
- present.n.01 -> present는 의미, n 은 명사, 01은 명사로서 가지는 의미를 구분하는 인덱스

-> synset 객체는 POS(품사), 정의, 부명제 등 시맨틱적인 요소를 표현

In [33]:
for synset in synsets:
    print('#### Synset name : ', synset.name(), '####')
    print(f'POS(품사) : {synset.lexname()}')
    print(f'Definition(정의) : {synset.definition()}')
    print(f'Lemmas(부명제) : {synset.lemma_names()}', '\n')

#### Synset name :  present.n.01 ####
POS(품사) : noun.time
Definition(정의) : the period of time that is happening now; any continuous stretch of time including the moment of speech
Lemmas(부명제) : ['present', 'nowadays'] 

#### Synset name :  present.n.02 ####
POS(품사) : noun.possession
Definition(정의) : something presented as a gift
Lemmas(부명제) : ['present'] 

#### Synset name :  present.n.03 ####
POS(품사) : noun.communication
Definition(정의) : a verb tense that expresses actions or states at the time of speaking
Lemmas(부명제) : ['present', 'present_tense'] 

#### Synset name :  show.v.01 ####
POS(품사) : verb.perception
Definition(정의) : give an exhibition of to an interested audience
Lemmas(부명제) : ['show', 'demo', 'exhibit', 'present', 'demonstrate'] 

#### Synset name :  present.v.02 ####
POS(품사) : verb.communication
Definition(정의) : bring forward and present to the mind
Lemmas(부명제) : ['present', 'represent', 'lay_out'] 

#### Synset name :  stage.v.01 ####
POS(품사) : verb.creation
Definition(정의

### WordNet : 어떤 어휘와 다른 어휘 간의 관계를 유사도로 나타낸 것

In [35]:
tree = wn.synset('tree.n.01')
lion = wn.synset('lion.n.01')
tiger = wn.synset('tiger.n.02')
cat = wn.synset('cat.n.01')
dog = wn.synset('dog.n.01')

entities = [tree, lion, tiger, cat, dog]
similarities = []
entity_names = [entity.name().split('.')[0] for entity in entities]

# 단어별 synset 을 반복하면서 다른 단어의 synset과 유사도를 측정
for entity in entities:
    similarity = [round(entity.path_similarity(compared_entity), 2)
                 for compared_entity in entities]
    similarities.append(similarity)
    
# 개별 단어별 synset과 다른 단어의 synset과의 유사도를 df로 저장
similarity_df = pd.DataFrame(similarities, columns = entity_names, index = entity_names)
similarity_df

tree  lion  tiger   cat   dog
tree   1.00  0.07   0.07  0.08  0.12
lion   0.07  1.00   0.33  0.25  0.17
tiger  0.07  0.33   1.00  0.25  0.17
cat    0.08  0.25   0.25  1.00  0.20
dog    0.12  0.17   0.17  0.20  1.00

### SentiWordNet : WordNet의 Synset과 유사한 클래스 (Senti_Synset 클래스를 리스트 형태로 반환)

In [37]:
import nltk
from nltk.corpus import sentiwordnet as swn

senti_synsets = list(swn.senti_synsets('slow'))
print('senti_synsets 반환 type: ', type(senti_synsets))
print('senti_synsets 반환 값 개수: ', len(senti_synsets))
print('senti_synsets 반환 type: ', senti_synsets)

senti_synsets 반환 type:  <class 'list'>
senti_synsets 반환 값 개수:  11
senti_synsets 반환 type:  [SentiSynset('decelerate.v.01'), SentiSynset('slow.v.02'), SentiSynset('slow.v.03'), SentiSynset('slow.a.01'), SentiSynset('slow.a.02'), SentiSynset('dense.s.04'), SentiSynset('slow.a.04'), SentiSynset('boring.s.01'), SentiSynset('dull.s.08'), SentiSynset('slowly.r.01'), SentiSynset('behind.r.03')]


- 단어의 감성을 나타내는 감성 지수와 객관성을 나타내는 객관성 지수를 가지고있음

- 감성 지수 : 긍정 감성 지수, 부정 감성 지수로 나뉨 (감성적이지 않으면 객관성 지수 :1, 감성 지수 : 0)

In [38]:
father = swn.senti_synset('father.n.01')
print('father 긍정감성 지수: ', father.pos_score())
print('father 부정감성 지수: ', father.neg_score())
print('father 객관성 지수: ', father.obj_score())

father 긍정감성 지수:  0.0
father 부정감성 지수:  0.0
father 객관성 지수:  1.0


In [39]:
fabulous = swn.senti_synset('fabulous.a.01')
print('fabulous  긍정감성 지수: ', fabulous .pos_score())
print('fabulous  부정감성 지수: ', fabulous .neg_score())
print('fabulous  객관성 지수: ', fabulous .obj_score())

fabulous  긍정감성 지수:  0.875
fabulous  부정감성 지수:  0.125
fabulous  객관성 지수:  0.0


#### 과제 1019_5. SentiWordNet 을 이용한 영화 감상평 감성분석을 수행

[감성 분석 순서]
1. 문서를 문장단위로 분해
2. 다시 문장을 단어 단위로 토큰화하고 품사 태깅
3. 품사 태깅된 단어 기반으로 synset 객체와 senti_synset 객체를 생성
4. Senti_synset에서 긍정 감성/부정 감성 지수를 구하고 이를 모두 합산해 특정 임계치 값 이상일 때 긍정 감성으로, 그렇지 않을 때는 부정 감성으로 결정

## [실습]: SentiWordNet 을 이용한 영화 감상평 감성분석을 수행

## 1. 감성 분석

In [ ]:
# WordNet을 이용해 문서를 단어로 토큰화한 후 어근 추출과 품사 태깅을 적용

In [43]:
# 품사 태깅을 수행하는 내부 함수 생성
from nltk.corpus import wordnet as wn

# NLTK PennTreebank Tag 기반으로 WordNet 기반의 품사 Tag로 변환
def penn_to_wn(tag):
    if tag.startswith('J'):
        return wn.ADJ
    elif tag.startswith('N'):
        return wn.NOUN
    elif tag.startswith('R'):
        return wn.ADV
    elif tag.startswith('V'):
        return wn.VERB

In [45]:
# 문장 -> 단어 토큰 -> 품사 태깅 후 SentiSynet클래스 생성, Polarity Score를 합산하는 함수 생성
from nltk.stem import WordNetLemmatizer
from nltk.corpus import sentiwordnet as swn
from nltk import sent_tokenize, word_tokenize, pos_tag

def swn_polarity(text):
    # 감성 지수 초기화
    sentiment = 0.0
    tokens_count = 0
    
    lemmatizer = WordNetLemmatizer() # 어근추출
    raw_sentences = sent_tokenize(text) # 문장으로 토큰화
    
    # 분해된 문장별로 단어 토큰 -> 품사 태깅 후 SentiSynset 생성 -> 감성 지수 합산
    for raw_sentence in raw_sentences:
        # NLTK 기반의 품사 태깅 문장 추출
        tagged_sentence = pos_tag(word_tokenize(raw_sentence)) #단어토큰화 후 품사 태깅
        for word, tag in tagged_sentence:
            
            #WordNet 기반 품사 태깅과 어근 추출(함수 적용)
            wn_tag = penn_to_wn(tag)
            if wn_tag not in (wn.NOUN, wn.ADJ, wn.ADV): # 동사는 제외 (명사, 형용사, 부사 나오지 않으면 다시 돌아감)
                continue
                
            lemma = lemmatizer.lemmatize(word, pos = wn_tag) # 해당 품사인 단어의 원형을 찾음
            if not lemma:
                continue
            
            # 어근을 추출한 단어와 WordNet기반 품사 태깅을 입력 -> Synset 객체 생성
            synsets = wn.synsets(lemma, pos = wn_tag)
            if not synsets:
                continue
                
            # sentiwordnet의 감성 단어 분석으로 감성 synset 추출
            # 긍정 감성 지수 : +, 부정 감성 지수 : - 로 합산
            
            synset = synsets[0]
            swn_synset = swn.senti_synset(synset.name())
            sentiment += (swn_synset.pos_score() - swn_synset.neg_score()) # 긍정지수 - 부정지수
            tokens_count += 1
            
    if not tokens_count:
        return 0
    
    # 총 score 가 0 이상일 경우 긍정 1, 그렇지 않을 경우 부정 0 반환
    
    if sentiment >= 0:
        return 1 # 긍정
    
    return 0 # 부정

## 2. 데이터 대입

### 1) 데이터 생성

In [47]:
# 데이터 생성

review_df = pd.read_csv('labeledTrainData.tsv', header=0, sep = '\t', quoting = 3)
review_df.head()

id  sentiment                                             review
0  "5814_8"          1  "With all this stuff going down at the moment ...
1  "2381_9"          1  "\"The Classic War of the Worlds\" by Timothy ...
2  "7759_3"          0  "The film starts with a manager (Nicholas Bell...
3  "3630_4"          0  "It must be assumed that those who praised thi...
4  "9495_8"          1  "Superbly trashy and wondrously unpretentious ...

In [48]:
import re

# < br >태그 공백 전환
review_df['review'] = review_df['review'].str.replace('<br />', ' ')

# 영어 문자열이 아닌 문자 전부 공백으로 전환
review_df['review'] = review_df['review'].apply(lambda x :re.sub("[^a-zA-Z]", " ", x))

review_df['review'][0]

' With all this stuff going down at the moment with MJ i ve started listening to his music  watching the odd documentary here and there  watched The Wiz and watched Moonwalker again  Maybe i just want to get a certain insight into this guy who i thought was really cool in the eighties just to maybe make up my mind whether he is guilty or innocent  Moonwalker is part biography  part feature film which i remember going to see at the cinema when it was originally released  Some of it has subtle messages about MJ s feeling towards the press and also the obvious message of drugs are bad m kay   Visually impressive but of course this is all about Michael Jackson so unless you remotely like MJ in anyway then you are going to hate this and find it boring  Some may call MJ an egotist for consenting to the making of this movie BUT MJ and most of his fans would say that he made it for the fans which if true is really nice of him   The actual feature film bit when it finally starts is only on for 

### 2) 데이터 대입

In [50]:
# 긍정인지 부정인지 예측해주는 함수 대입
review_df['preds'] = review_df['review'].apply(lambda x : swn_polarity(x))

# 실제 강성 평가를 타깃값으로 
y_target = review_df['sentiment'].values

# 함수를 대입한 값을 예측값으로
preds = review_df['preds'].values

In [52]:
# 예측값, 실제 감성 평가 결과만 봐보기
review_df[['sentiment', 'preds']]

sentiment  preds
0              1      0
1              1      1
2              0      0
3              0      0
4              1      0
...          ...    ...
24995          0      0
24996          0      1
24997          0      1
24998          0      0
24999          1      0

[25000 rows x 2 columns]

## 3. 예측 성능 평가

In [65]:
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, f1_score, roc_auc_score

print( '오차행렬: ', '\n', confusion_matrix(y_target, preds))
print(f'정확도 : {accuracy_score(y_target, preds):.3f}')
print(f'정밀도 : {precision_score(y_target, preds):.3f}')
print(f'재현율 : {recall_score(y_target, preds):.3f}')
print(f'F1 : {f1_score(y_target, preds):.3f}')

오차행렬:  
 [[7668 4832]
 [3636 8864]]
정확도 : 0.661
정밀도 : 0.647
재현율 : 0.709
F1 : 0.677


## 2) VADER를 이용한 감성 분석

소셜 미디어의 감성 분석 용도로 만들어진 룰 기반의 Lexicon

## [실습]: VADER를 이용한 영화 감상평 감성 분석

## 1. 데이터 생성

In [104]:
# 데이터 생성

review_df = pd.read_csv('labeledTrainData.tsv', header=0, sep = '\t', quoting = 3)

import re

# < br >태그 공백 전환
review_df['review'] = review_df['review'].str.replace('<br />', ' ')

# 영어 문자열이 아닌 문자 전부 공백으로 전환
review_df['review'] = review_df['review'].apply(lambda x :re.sub("[^a-zA-Z]", " ", x))

review_df['review'][0]

' With all this stuff going down at the moment with MJ i ve started listening to his music  watching the odd documentary here and there  watched The Wiz and watched Moonwalker again  Maybe i just want to get a certain insight into this guy who i thought was really cool in the eighties just to maybe make up my mind whether he is guilty or innocent  Moonwalker is part biography  part feature film which i remember going to see at the cinema when it was originally released  Some of it has subtle messages about MJ s feeling towards the press and also the obvious message of drugs are bad m kay   Visually impressive but of course this is all about Michael Jackson so unless you remotely like MJ in anyway then you are going to hate this and find it boring  Some may call MJ an egotist for consenting to the making of this movie BUT MJ and most of his fans would say that he made it for the fans which if true is really nice of him   The actual feature film bit when it finally starts is only on for 

### (예시) 감상평 한 개만 VADER를 이용해 감성 분석

In [105]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer

senti_analyzer = SentimentIntensityAnalyzer() # 객체 생성
senti_scores = senti_analyzer.polarity_scores(review_df['review'][0]) # polarity_scores : 감성 점수 추출 매서드(neg : 부정 감성 지수, neu : 중립적인 감성 지수, pos : 긍정 감성 지수, compund : neg, neu, pos를 적절히 조합해 -1~1사이의 감성 지수를 표현한 값)
print(senti_scores)

{'neg': 0.13, 'neu': 0.743, 'pos': 0.127, 'compound': -0.7943}


[polarity_scores 매서드]

추출값

neg : 부정 감성 지수, neu : 중립적인 감성 지수, pos : 긍정 감성 지수, compund : neg, neu, pos를 적절히 조합해 -1~1사이의 감성 지수를 표현한 값

-> compund 값으로 부정 감성 또는 긍정 감성 여부를 결정 (주로 0.1이상이면 긍정 감성) _ 임계값 조정 가능

In [107]:
analyzer = SentimentIntensityAnalyzer()
scores = analyzer.polarity_scores(review_df['review'][0])
scores

{'neg': 0.13, 'neu': 0.743, 'pos': 0.127, 'compound': -0.7943}

In [108]:
scores['compound']

-0.7943

In [109]:
# 감상평, 임계값을 입력하면 compound 값을 기반으로 긍정인지 부정인지 예측 하는 함수 생성

def vader_polarity(review, threshold = 0.1):
    analyzer = SentimentIntensityAnalyzer()
    scores = analyzer.polarity_scores(review) # 감상평에 감성 점수 추출 
    
    # compund 값이 임계값보다 크면 긍정, 아니면 부정으로 인식
    agg_score = scores['compound']
    final_sentiment = 1 if agg_score >= threshold else 0
    return final_sentiment

# 레코드별로 함수 수행
review_df['vader_preds'] = review_df['review'].apply(lambda x : vader_polarity(x, 0.1)) #임계값 
y_target = review_df['sentiment'].values
preds = review_df['vader_preds'].values

review_df[['sentiment', 'vader_preds']]

sentiment  vader_preds
0              1            0
1              1            1
2              0            0
3              0            1
4              1            1
...          ...          ...
24995          0            1
24996          0            1
24997          0            1
24998          0            0
24999          1            0

[25000 rows x 2 columns]

In [110]:
print( '오차행렬: ', '\n', confusion_matrix(y_target, preds))
print(f'정확도 : {accuracy_score(y_target, preds):.3f}')
print(f'정밀도 : {precision_score(y_target, preds):.3f}')
print(f'재현율 : {recall_score(y_target, preds):.3f}')
print(f'F1 : {f1_score(y_target, preds):.3f}')

오차행렬:  
 [[ 6747  5753]
 [ 1858 10642]]
정확도 : 0.696
정밀도 : 0.649
재현율 : 0.851
F1 : 0.737


[SentiWordNet 의 감성 분석 결과]

오차행렬:  
 [[7668 4832]
 [3636 8864]]
정확도 : 0.661
정밀도 : 0.647
재현율 : 0.709
F1 : 0.677

VADER 을 활용한 분석 결과가 더 정확